File: backend/app/languages.py
from __future__ import annotations

# Shared language registry for the backend.
#
# The frontend sends human-readable names such as "Hindi"; the backend converts
# them to the model-specific IndicTrans codes below before calling translation.
# If a language is added here, check the frontend language list and TTS/ASR
# fallback maps too.

# Canonical language names mapped to IndicTrans language codes.
LANGUAGE_TO_CODE = {
    "English": "eng_Latn",
    "Assamese": "asm_Beng",
    "Bodo": "brx_Deva",
    "Dogri": "doi_Deva",
    "Gujarati": "guj_Gujr",
    "Hindi": "hin_Deva",
    "Kannada": "kan_Knda",
    "Kashmiri": "kas_Arab",
    "Konkani": "gom_Deva",
    "Maithili": "mai_Deva",
    "Malayalam": "mal_Mlym",
    "Bengali": "ben_Beng",
    "Manipuri": "mni_Mtei",
    "Marathi": "mar_Deva",
    "Nepali": "npi_Deva",
    "Odia": "ory_Orya",
    "Punjabi": "pan_Guru",
    "Sanskrit": "san_Deva",
    "Santali": "sat_Olck",
    "Sindhi": "snd_Arab",
    "Tamil": "tam_Taml",
    "Telugu": "tel_Telu",
    "Urdu": "urd_Arab",
}

# Dialect and alternate names that are normalized to canonical names above.
# These aliases let users or API clients pass familiar names while the model
# still receives one supported canonical language.
LANGUAGE_ALIASES = {
    # Hindi cluster dialects.
    "Awadhi": "Hindi",
    "Avadhi": "Hindi",
    "Bhojpuri": "Hindi",
    "Braj": "Hindi",
    "Bundeli": "Hindi",
    "Chhattisgarhi": "Hindi",
    "Garhwali": "Hindi",
    "Haryanvi": "Hindi",
    "Kumaoni": "Hindi",
    "Magahi": "Hindi",
    "Marwari": "Hindi",
    # Bengali cluster dialects.
    "Sylheti": "Bengali",
    "Chittagonian": "Bengali",
    # Urdu variants.
    "Dakhini": "Urdu",
    "Hyderabadi Urdu": "Urdu",
    # Kannada variants.
    "Tulu": "Kannada",
    "Kodava": "Kannada",
    # Common alternate spellings.
    "Bangla": "Bengali",
    "Oriya": "Odia",
    "Meitei": "Manipuri",
}

File: Expo/constants/languages.ts
// Language names used by the Expo app.
//
// These strings are sent to the backend, which maps them to model-specific
// language codes in backend/app/languages.py. Keep both files synchronized when
// adding, removing, or renaming a language.

export const SUPPORTED_LANGUAGES = [
  'English',
  'Assamese',
  'Bodo',
  'Dogri',
  'Gujarati',
  'Hindi',
  'Kannada',
  'Kashmiri',
  'Konkani',
  'Maithili',
  'Malayalam',
  'Bengali',
  'Manipuri',
  'Marathi',
  'Nepali',
  'Odia',
  'Punjabi',
  'Sanskrit',
  'Santali',
  'Sindhi',
  'Tamil',
  'Telugu',
  'Urdu',
] as const;

export type SupportedLanguage = (typeof SUPPORTED_LANGUAGES)[number];
const SUPPORTED_LANGUAGE_SET = new Set<string>(SUPPORTED_LANGUAGES);

// Labels shown in the picker. The object keys must stay exactly equal to the
// SupportedLanguage names above because getLanguageLabel indexes this map.
export const LANGUAGE_LABELS: Record<SupportedLanguage, string> = {
  English: 'अंग्रेज़ी',
  Assamese: 'অসমীয়া',
  Bodo: 'बड़ो',
  Dogri: 'डोगरी',
  Gujarati: 'ગુજરાતી',
  Hindi: 'हिन्दी',
  Kannada: 'ಕನ್ನಡ',
  Kashmiri: 'कश्मीरी',
  Konkani: 'कोंकणी',
  Maithili: 'मैथिली',
  Malayalam: 'മലയാളം',
  Bengali: 'বাংলা',
  Manipuri: 'মণিপুরী',
  Marathi: 'मराठी',
  Nepali: 'नेपाली',
  Odia: 'ଓଡ଼ିଆ',
  Punjabi: 'ਪੰਜਾਬੀ',
  Sanskrit: 'संस्कृत',
  Santali: 'संताली',
  Sindhi: 'सिंधी',
  Tamil: 'தமிழ்',
  Telugu: 'తెలుగు',
  Urdu: 'اردو',
};

export function isSupportedLanguage(language: string): language is SupportedLanguage {
  return SUPPORTED_LANGUAGE_SET.has(language);
}

export function getLanguageLabel(language: SupportedLanguage | string): string {
  return isSupportedLanguage(language) ? LANGUAGE_LABELS[language] : language;
}
